src/grain/ledger.js

```diff
@@ -0,0 +1,109 @@
+// @flow
+
+import {type NodeAddressT, NodeAddress} from "../core/graph";
+import stringify from "json-stable-stringify";
+import * as NullUtil from "../util/null";
+import {type Grain, format as formatGrain} from "./grain";
+
+export type GrainDistribution = {|
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type GrainTransfer = {|
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type AliasSpec = {|
+  +canonical: NodeAddressT,
+  +alias: NodeAddressT,
+|};
+
+export type LedgerHistory = {|
+  +transfers: $ReadOnlyArray<GrainTransfer>,
+  +distributions: $ReadOnlyArray<GrainDistribution>,
+  +aliases: Map<NodeAddressT, NodeAddressT>,
+|};
+
+export interface Ledger {
+  balances(): Map<NodeAddressT, Grain>;
+  earnings(): Map<NodeAddressT, Grain>;
```

Comment:
Earnings is a particularly useful thing to cache because when computing distributions, we need to know the lifetime earnings for every contributor--that way we can compare their lifetime payout to their total cred.